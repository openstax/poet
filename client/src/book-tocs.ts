import { EventEmitter, type TreeItem, TreeItemCollapsibleState, Uri, type TreeDataProvider } from 'vscode'

import { type BookToc, type ClientTocNode, BookRootNode, TocNodeKind, type ClientPageish } from '../../common/src/toc'
import { TocItemIcon } from './toc-trees-provider'

export type BookOrTocNode = BookToc | ClientTocNode

const toClientTocNode = (n: ClientPageish): ClientTocNode => ({ type: TocNodeKind.Page, value: n })

export class TocsTreeProvider implements TreeDataProvider<BookOrTocNode> {
  private readonly _onDidChangeTreeData = new EventEmitter<void>()
  readonly onDidChangeTreeData = this._onDidChangeTreeData.event

  public includeFileIdsForFilter = false
  private bookTocs: BookToc[]
  private orphans: BookOrTocNode[]
  private readonly parentsMap = new Map<BookOrTocNode, BookOrTocNode>()

  constructor() {
    this.bookTocs = []
    this.orphans = []
  }

  public toggleFilterMode() {
    this.includeFileIdsForFilter = !this.includeFileIdsForFilter
    this._onDidChangeTreeData.fire()
  }

  public update(n: BookToc[], o: ClientPageish[]) {
    this.bookTocs = n
    this.parentsMap.clear()
    this.bookTocs.forEach(n => { this.recAddParent(n) })
    this.orphans = o.map(toClientTocNode)
    this._onDidChangeTreeData.fire()
  }

  private recAddParent(node: BookOrTocNode) {
    const kids = this.getChildren(node)
    kids.forEach(k => {
      this.parentsMap.set(k, node)
      this.recAddParent(k)
    })
  }

  public getTreeItem(node: BookOrTocNode): TreeItem {
    if (node.type === BookRootNode.Singleton) {
      const uri = Uri.parse(node.absPath)
      return {
        iconPath: TocItemIcon.Book,
        collapsibleState: TreeItemCollapsibleState.Collapsed,
        label: node.title,
        description: node.slug,
        resourceUri: uri,
        command: { title: 'open', command: 'vscode.open', arguments: [uri] }
      }
    } else if (node.type === TocNodeKind.Page) {
      const uri = Uri.parse(node.value.absPath)
      const title = node.value.title ?? 'Loading...'
      const ret = this.includeFileIdsForFilter ? { label: `${title} (${node.value.fileId})` } : { label: title, description: node.value.fileId }
      return {
        ...ret,
        iconPath: TocItemIcon.Page,
        collapsibleState: TreeItemCollapsibleState.None,
        resourceUri: uri,
        command: { title: 'open', command: 'vscode.open', arguments: [uri] },
        contextValue: TocNodeKind.Page
      }
    } else {
      return {
        iconPath: TocItemIcon.Subbook,
        collapsibleState: TreeItemCollapsibleState.Collapsed,
        label: node.value.title,
        contextValue: TocNodeKind.Subbook
      }
    }
  }

  public getChildren(node?: BookOrTocNode) {
    let kids: BookOrTocNode[] = []
    if (node === undefined) {
      return [...this.bookTocs, ...this.orphans]
    } else if (node.type === BookRootNode.Singleton) {
      kids = node.tocTree
    } else if (node.type === TocNodeKind.Page) {
      kids = []
    } else {
      kids = node.children
    }
    return kids
  }

  public getParent(node: BookOrTocNode) {
    return this.parentsMap.get(node)
  }

  public getParentBook(node: BookOrTocNode): BookToc | undefined {
    let bookNode: BookOrTocNode | undefined = node
    // Books do not have parent books
    if (bookNode.type === BookRootNode.Singleton) return undefined
    while (bookNode !== undefined && bookNode.type !== BookRootNode.Singleton) {
      bookNode = this.getParent(bookNode)
    }
    return bookNode?.type === BookRootNode.Singleton ? bookNode : undefined
  }

  public getParentBookIndex(node: BookOrTocNode) {
    const parentBook = this.getParentBook(node)
    let idx: number | undefined
    if (parentBook !== undefined) {
      idx = this.bookTocs.findIndex((b) => b.absPath === parentBook.absPath)
    }
    return idx === -1 ? undefined : idx
  }
}
