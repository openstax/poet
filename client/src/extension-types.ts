/* istanbul ignore file */
export enum PanelType {
  TOC_EDITOR = 'openstax.tocEditor',
  IMAGE_MANAGER = 'openstax.imageManager',
  CNXML_PREVIEW = 'openstax.cnxmlPreview'
}
export enum OpenstaxCommand {
  SHOW_TOC_EDITOR = 'openstax.showTocEditor',
  SHOW_CNXML_PREVIEW = 'openstax.showPreviewToSide',
  SHOW_IMAGE_MANAGER = 'openstax.showImageManager'
}
