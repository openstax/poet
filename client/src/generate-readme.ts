import vscode from 'vscode'
import { requestGenerateReadme } from '../../common/src/requests'
import { ExtensionHostContext } from './panel'
import { expect, getRootPathUri } from './utils'

export function readmeGenerator(hostContext: ExtensionHostContext) {
  return async () => {
    await requestGenerateReadme(
      hostContext.client, {
        workspaceUri: expect(getRootPathUri(), 'Could not get root workspace uri').toString()
      }
    )
    void vscode.window.showInformationMessage('Done!', { title: 'Generate README' })
  }
}
