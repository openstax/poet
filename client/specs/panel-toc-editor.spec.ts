import { join } from 'path'
import { expect } from '@jest/globals'
import SinonRoot, { SinonStub } from 'sinon'

import vscode, { Disposable, Event, EventEmitter, Uri, ViewColumn, WebviewPanel } from 'vscode'
import { BookRootNode, BookToc, TocNodeKind, TocModificationKind } from '../../common/src/toc'
import * as utils from '../src/utils' // Used for dependency mocking in tests
import { TocItemIcon, TocTreeItem, TocTreesProvider, toggleTocTreesFilteringHandler } from '../src/toc-trees-provider'
import { PanelIncomingMessage, TocEditorPanel } from '../src/panel-toc-editor'
import { LanguageClient } from 'vscode-languageclient/node'
import { EMPTY_BOOKS_AND_ORPHANS, ExtensionServerRequest } from '../../common/src/requests'
import { ExtensionEvents, ExtensionHostContext } from '../src/panel'
import { BookOrTocNode, TocsTreeProvider } from '../src/book-tocs'
import { PanelStateMessage, PanelStateMessageType } from '../../common/src/webview-constants'

const TEST_OUT_DIR = join(__dirname, '../src')
const resourceRootDir = TEST_OUT_DIR

const createMockClient = () => {
  const sendRequestStub = SinonRoot.stub()
  sendRequestStub.returns([])
  const client = {
    sendRequest: sendRequestStub,
    onRequest: SinonRoot.stub().returns({ dispose: () => { } })
  } as unknown as LanguageClient

  return {
    client,
    sendRequestStub
  }
}

type ExtractEventGeneric<GenericEvent> = GenericEvent extends Event<infer X> ? X : never
type ExtensionEventEmitters = { [key in keyof ExtensionEvents]: EventEmitter<ExtractEventGeneric<ExtensionEvents[key]>> }
const createMockEvents = (): { emitters: ExtensionEventEmitters, events: ExtensionEvents } => {
  const onDidChangeWatchedFilesEmitter: EventEmitter<undefined> = new EventEmitter()
  const emitters = {
    onDidChangeWatchedFiles: onDidChangeWatchedFilesEmitter
  }
  const events = {
    onDidChangeWatchedFiles: onDidChangeWatchedFilesEmitter.event
  }
  return { emitters, events }
}

describe('Toc Editor', () => {
  const sinon = SinonRoot.createSandbox()
  afterEach(() => sinon.restore())
  it('TocTreesProvider returns expected TocTreeItems', async () => {
    const fakeTreeBooks: BookToc[] = []
    fakeTreeBooks.push(
      {
        type: BookRootNode.Singleton,
        title: 'Book1',
        slug: 'slug1',
        uuid: '',
        language: '',
        licenseUrl: '',
        absPath: 'path/to/nowhere-book',
        tocTree: [{
          type: TocNodeKind.Subbook,
          value: { token: 'id123', title: 'subbook' },
          children: [{
            type: TocNodeKind.Page,
            value: {
              token: 'id234',
              absPath: 'path/to/nowhere',
              fileId: 'm00001',
              title: 'Module1'
            }
          },
          {
            type: TocNodeKind.Page,
            value: {
              token: 'id345',
              absPath: 'path/to/nowhere2',
              fileId: 'm00002',
              title: 'Module2'
            }
          }]
        }]
      },
      {
        type: BookRootNode.Singleton,
        title: 'Book2',
        slug: 'slug2',
        uuid: '',
        language: '',
        licenseUrl: '',
        absPath: 'path/to/nowhere-book',
        tocTree: [{
          type: TocNodeKind.Page,
          value: {
            token: 'id123',
            absPath: 'path/to/nowhere3',
            fileId: 'm00003',
            title: 'Module3'
          }
        }]
      }
    )
    const fakeWorkspacePath = '/tmp/fakeworkspace'
    sinon.stub(utils, 'getRootPathUri').returns(vscode.Uri.file(fakeWorkspacePath))
    const module1Item = new TocTreeItem(
      TocItemIcon.Page,
      'Module1',
      vscode.TreeItemCollapsibleState.None,
      [],
      {
        title: 'open',
        command: 'vscode.open',
        arguments: [vscode.Uri.file(`${fakeWorkspacePath}/modules/m00001/index.cnxml`)]
      },
      'm00001'
    )
    const module2Item = new TocTreeItem(
      TocItemIcon.Page,
      'Module2',
      vscode.TreeItemCollapsibleState.None,
      [],
      {
        title: 'open',
        command: 'vscode.open',
        arguments: [vscode.Uri.file(`${fakeWorkspacePath}/modules/m00002/index.cnxml`)]
      },
      'm00002'
    )
    const module3Item = new TocTreeItem(
      TocItemIcon.Page,
      'Module3',
      vscode.TreeItemCollapsibleState.None,
      [],
      {
        title: 'open',
        command: 'vscode.open',
        arguments: [vscode.Uri.file(`${fakeWorkspacePath}/modules/m00003/index.cnxml`)]
      },
      'm00003'
    )
    const subbookItem = new TocTreeItem(
      TocItemIcon.Subbook,
      'subbook',
      vscode.TreeItemCollapsibleState.Collapsed,
      [module1Item, module2Item]
    )
    const book1Item = new TocTreeItem(
      TocItemIcon.Book,
      'Book1',
      vscode.TreeItemCollapsibleState.Collapsed,
      [subbookItem],
      {
        title: 'open',
        command: 'vscode.open',
        arguments: [vscode.Uri.file(`${fakeWorkspacePath}/collections/slug1.collection.xml`)]
      }
    )
    const book2Item = new TocTreeItem(
      TocItemIcon.Book,
      'Book2',
      vscode.TreeItemCollapsibleState.Collapsed,
      [module3Item],
      {
        title: 'open',
        command: 'vscode.open',
        arguments: [vscode.Uri.file(`${fakeWorkspacePath}/collections/slug2.collection.xml`)]
      }
    )

    const mockClient = createMockClient().client
    // We don't want to just return []
    const sendRequestMock = sinon.stub()
    mockClient.sendRequest = sendRequestMock
    const tocTreesProvider = new TocTreesProvider({ bookTocs: EMPTY_BOOKS_AND_ORPHANS, resourceRootDir, client: mockClient, events: createMockEvents().events })
    sendRequestMock.onCall(0).resolves(null)
    sendRequestMock.onCall(1).resolves(fakeTreeBooks)

    expect(await tocTreesProvider.getChildren(undefined)).toMatchSnapshot()
    expect(await tocTreesProvider.getChildren(undefined)).toMatchSnapshot()
    expect(book1Item).toMatchSnapshot()
    expect(await tocTreesProvider.getChildren(book2Item)).toMatchSnapshot()
    expect(tocTreesProvider.getTreeItem(book2Item)).toMatchSnapshot()
    expect(await tocTreesProvider.getParent(book2Item)).toMatchSnapshot()
    expect(await tocTreesProvider.getParent(module3Item)).toMatchSnapshot()
    expect(await tocTreesProvider.getParent(module1Item)).toMatchSnapshot()
    tocTreesProvider.toggleFilterMode()
    expect(tocTreesProvider.getTreeItem(module3Item)).toMatchSnapshot()
  })

  describe('PanelTocEditor', () => {
    let postMessageStub = undefined as unknown as SinonStub<[message: any], Thenable<boolean>>
    let onDidReceiveMessageStub: SinonRoot.SinonStub<[listener: (e: any) => any, thisArgs?: any, disposables?: vscode.Disposable[] | undefined], vscode.Disposable>
    let watchedFilesSpy: SinonRoot.SinonSpy<[listener: (e: any) => any, thisArgs?: any, disposables?: vscode.Disposable[]], vscode.Disposable>
    let p = undefined as unknown as TocEditorPanel
    const { client, sendRequestStub } = createMockClient()
    const { emitters, events } = createMockEvents()
    const context: ExtensionHostContext = {
      resourceRootDir: join(__dirname, '..', 'static'), // HTML webview files are loaded from here
      client,
      events,
      bookTocs: { books: [], orphans: [] }
    }
    beforeEach(() => {
      const webviewPanel = vscode.window.createWebviewPanel('unused', 'unused', ViewColumn.Active)
      postMessageStub = sinon.stub(webviewPanel.webview, 'postMessage')
      onDidReceiveMessageStub = sinon.stub(webviewPanel.webview, 'onDidReceiveMessage')
      onDidReceiveMessageStub.returns(new Disposable(sinon.stub()))

      watchedFilesSpy = sinon.spy(events, 'onDidChangeWatchedFiles')
      sinon.stub(vscode.window, 'createWebviewPanel').returns(webviewPanel)

      p = new TocEditorPanel(context)
    })
    it('calls handleMessage when the Webview sends a message', () => {
      const handleMessageStub = sinon.stub(p, 'handleMessage').returns(Promise.resolve())
      const message: PanelIncomingMessage = {
        type: TocModificationKind.Remove,
        nodeToken: 'my-token-id',
        bookIndex: 0
      }
      expect(handleMessageStub.callCount).toBe(0)
      expect(onDidReceiveMessageStub.callCount).toBe(1)
      const cb = onDidReceiveMessageStub.firstCall.args[0]
      cb(message)
      expect(handleMessageStub.callCount).toBe(1)
    })
    it('translates events from the webview and sends them to the language server', async () => {
      let callCount = 0
      function getMessage() {
        const reqType = ExtensionServerRequest.TocModification
        if (sendRequestStub.callCount <= callCount) { throw new Error('expected sendRequest to have been called but it was not') }
        const c = sendRequestStub.getCall(callCount++)
        if (c.firstArg !== reqType) { throw new Error(`expected the first arg of sendRequest to be '${reqType}' but it was '${c.firstArg as unknown as string}'`) }
        return c.args[1].event
      }
      const uberEvent = {
        newTitle: 'my_new_title',
        nodeToken: 'mytoken',
        newParentToken: undefined,
        newChildIndex: 0,
        bookIndex: 0
      }

      sinon.stub(vscode.workspace, 'workspaceFolders').get(() => [{ uri: Uri.file('/path/to/workspace/root') }])
      await p.handleMessage({ ...uberEvent, type: TocModificationKind.Move })
      expect(getMessage().type).toBe(TocModificationKind.Move)
      await p.handleMessage({ ...uberEvent, type: TocModificationKind.Remove })
      expect(getMessage().type).toBe(TocModificationKind.Remove)
      await p.handleMessage({ ...uberEvent, type: TocModificationKind.PageRename })
      expect(getMessage().type).toBe(TocModificationKind.PageRename)
      await p.handleMessage({ ...uberEvent, type: TocModificationKind.SubbookRename })
      expect(getMessage().type).toBe(TocModificationKind.SubbookRename)

      sinon.stub(vscode.window, 'showInputBox').returns(Promise.resolve('new_title'))
      await p.handleMessage({ type: TocNodeKind.Page, title: 'foo', bookIndex: 0 })
      expect(getMessage().title).toBe('new_title')

      await p.handleMessage({ type: TocNodeKind.Subbook, title: 'foo', bookIndex: 0, slug: 'subbook_slug' })
      expect(getMessage().title).toBe('new_title')
    })
    it('disposes', () => {
      expect(() => p.dispose()).not.toThrow()
    })
    it('sends a message to Webview when a fileChanged event is emitted', () => {
      expect(postMessageStub.callCount).toBe(0)
      emitters.onDidChangeWatchedFiles.fire()
      expect(postMessageStub.callCount).toBe(1)
    })
    it('sends a message to Webview when the content is updated', async () => {
      const testToc: BookToc = {
        type: BookRootNode.Singleton,
        absPath: '/fake/path',
        uuid: 'uuid',
        title: 'title',
        slug: 'slug',
        language: 'language',
        licenseUrl: 'licenseUrl',
        tocTree: [{
          type: TocNodeKind.Subbook,
          value: {
            token: 'token',
            title: 'title'
          },
          children: [{
            type: TocNodeKind.Page,
            value: {
              token: 'token',
              title: 'title',
              fileId: 'fileId',
              absPath: '/fake/path/to/file'
            }
          }]
        }]
      }
      const v1 = { books: [testToc], orphans: [] }

      expect(postMessageStub.callCount).toBe(0)
      await p.update(v1)
      expect(postMessageStub.callCount).toBe(1)
      expect(postMessageStub.firstCall.args).toMatchSnapshot()
    })
    it('does not send a message to Webview when panel is disposed', async () => {
      await expect(p.refreshPanel({} as unknown as WebviewPanel, client)).rejects
    })
    it('refreshes when server watched file changes', async () => {
      const refreshStub = sinon.stub(p, 'refreshPanel')
      await watchedFilesSpy.getCall(0).args[0](undefined)
      expect(refreshStub.called).toBe(true)
    })
    it('sorts pages based on fileid', async () => {
      const testToc: BookToc = {
        type: BookRootNode.Singleton,
        absPath: '/fake/path',
        uuid: 'uuid',
        title: 'title',
        slug: 'slug',
        language: 'language',
        licenseUrl: 'licenseUrl',
        tocTree: [{
          type: TocNodeKind.Page,
          value: {
            token: 'token',
            title: 'title',
            fileId: 'fileId2',
            absPath: '/fake/path/to/file2'
          }
        }, {
          type: TocNodeKind.Page,
          value: {
            token: 'token',
            title: 'title',
            fileId: 'fileId1',
            absPath: '/fake/path/to/file1'
          }
        }]
      }
      const v = { books: [testToc], orphans: [] }
      expect(postMessageStub.callCount).toBe(0)
      await p.update(v)
      const message: PanelStateMessage<any> = postMessageStub.firstCall.args[0]
      expect(message.type).toBe(PanelStateMessageType.Response)
      const allModules = message.state.uneditable[0]
      expect(allModules.tocTree.length).toBe(2)
      expect(allModules.tocTree[0].fileId).toBe('fileId1')
      expect(allModules.tocTree[1].fileId).toBe('fileId2')
    })
  })

  describe('filtering', () => {
    it('toggleTocTreesFilteringHandler', async () => {
      const revealStub = sinon.stub()
      const toggleFilterStub = sinon.stub()
      const getChildrenStub = sinon.stub()
      const refreshStub = sinon.stub()

      const view: vscode.TreeView<BookOrTocNode> = {
        reveal: revealStub
      } as unknown as vscode.TreeView<BookOrTocNode>
      const provider: TocsTreeProvider = {
        toggleFilterMode: toggleFilterStub,
        getChildren: getChildrenStub,
        refresh: refreshStub,
        getParent: () => undefined
      } as unknown as TocsTreeProvider
      const fakeChildren = [
        { type: BookRootNode.Singleton, tocTree: [{ type: TocNodeKind.Subbook, label: 'unit1', children: [{ type: TocNodeKind.Subbook, label: 'subcol1', children: [{ type: TocNodeKind.Page, label: 'm2', children: [] }] }] }] },
        { type: BookRootNode.Singleton, tocTree: [{ label: 'm1', children: [] }] }
      ]
      getChildrenStub.returns(fakeChildren)

      const handler = toggleTocTreesFilteringHandler(view, provider)
      await handler()
      expect(toggleFilterStub.callCount).toBe(1)
      expect(getChildrenStub.callCount).toBe(1)
      expect(revealStub.callCount).toBe(2)
      expect(revealStub.getCalls().map(c => c.args)).toMatchSnapshot()
      expect(refreshStub.callCount).toBe(0)
    })
    it('toggleTocTreesFilteringHandler disables itself while revealing', async () => {
      const revealStub = sinon.stub()
      const toggleFilterStub = sinon.stub()
      const getChildrenStub = sinon.stub()
      const fakeChildren = [
        { label: 'col1', children: [{ label: 'm1', children: [] }] }
      ]
      getChildrenStub.returns(fakeChildren)

      const view: vscode.TreeView<BookOrTocNode> = {
        reveal: revealStub
      } as unknown as vscode.TreeView<BookOrTocNode>
      const provider: TocsTreeProvider = {
        toggleFilterMode: toggleFilterStub,
        getChildren: getChildrenStub,
        getParent: () => undefined
      } as unknown as TocsTreeProvider

      const handler = toggleTocTreesFilteringHandler(view, provider)
      // Invoke the handler the first time reveal is called to simulate a parallel
      // user request without resorting to synthetic delay injection
      revealStub.onCall(0).callsFake(handler)
      await handler()
      expect(toggleFilterStub.callCount).toBe(1)
      expect(revealStub.callCount).toBe(1)
      expect(getChildrenStub.callCount).toBe(1)
    })
    it('toggleTocTreesFilteringHandler does not lock itself on errors', async () => {
      const toggleFilterStub = sinon.stub().throws()
      const view: vscode.TreeView<BookOrTocNode> = {} as unknown as vscode.TreeView<BookOrTocNode>
      const provider: TocsTreeProvider = {
        toggleFilterMode: toggleFilterStub
      } as unknown as TocsTreeProvider

      const handler = toggleTocTreesFilteringHandler(view, provider)
      try { await handler() } catch { }
      try { await handler() } catch { }
      expect(toggleFilterStub.callCount).toBe(2)
    })
  })
})
