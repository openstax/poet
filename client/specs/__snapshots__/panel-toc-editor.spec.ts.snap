// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toc Editor PanelTocEditor sends a message to Webview when the content is updated 1`] = `
Array [
  Object {
    "editable": Array [
      Object {
        "absPath": "/fake/path",
        "language": "language",
        "licenseUrl": "licenseUrl",
        "slug": "slug",
        "title": "title",
        "tree": Array [
          Object {
            "children": Array [
              Object {
                "absPath": "/fake/path/to/file",
                "fileId": "fileId",
                "subtitle": "fileId",
                "title": "title",
                "token": "token",
                "type": "TocNodeKind.Leaf",
              },
            ],
            "title": "title",
            "token": "token",
            "type": "TocNodeKind.Inner",
          },
        ],
        "type": "BookRootNode.Singleton",
        "uuid": "uuid",
      },
    ],
    "uneditable": Array [
      Object {
        "slug": "mock-slug__source-only",
        "title": "All Modules",
        "tree": Array [
          Object {
            "absPath": "/fake/path/to/file",
            "fileId": "fileId",
            "subtitle": "fileId",
            "title": "title",
            "token": "token",
            "type": "TocNodeKind.Leaf",
          },
        ],
      },
      Object {
        "slug": "mock-slug__source-only",
        "title": "Orphan Modules",
        "tree": Array [],
      },
    ],
  },
]
`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 1`] = `Array []`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 2`] = `Array []`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 3`] = `
TocTreeItem {
  "children": Array [
    TocTreeItem {
      "children": Array [
        TocTreeItem {
          "children": Array [],
          "collapsibleState": "None",
          "command": Object {
            "arguments": Array [
              Object {
                "$mid": 1,
                "path": "/tmp/fakeworkspace/modules/m00001/index.cnxml",
                "scheme": "file",
              },
            ],
            "command": "vscode.open",
            "title": "open",
          },
          "description": "m00001",
          "iconPath": "File",
          "label": "Module1",
          "parent": [Circular],
        },
        TocTreeItem {
          "children": Array [],
          "collapsibleState": "None",
          "command": Object {
            "arguments": Array [
              Object {
                "$mid": 1,
                "path": "/tmp/fakeworkspace/modules/m00002/index.cnxml",
                "scheme": "file",
              },
            ],
            "command": "vscode.open",
            "title": "open",
          },
          "description": "m00002",
          "iconPath": "File",
          "label": "Module2",
          "parent": [Circular],
        },
      ],
      "collapsibleState": "Collapsed",
      "command": undefined,
      "description": undefined,
      "iconPath": "Folder",
      "label": "subcollection",
      "parent": [Circular],
    },
  ],
  "collapsibleState": "Collapsed",
  "command": Object {
    "arguments": Array [
      Object {
        "$mid": 1,
        "path": "/tmp/fakeworkspace/collections/collection1.collection.xml",
        "scheme": "file",
      },
    ],
    "command": "vscode.open",
    "title": "open",
  },
  "description": undefined,
  "iconPath": ThemeIcon {},
  "label": "Collection1",
  "parent": undefined,
}
`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 4`] = `
Array [
  TocTreeItem {
    "children": Array [],
    "collapsibleState": "None",
    "command": Object {
      "arguments": Array [
        Object {
          "$mid": 1,
          "path": "/tmp/fakeworkspace/modules/m00003/index.cnxml",
          "scheme": "file",
        },
      ],
      "command": "vscode.open",
      "title": "open",
    },
    "description": "m00003",
    "iconPath": "File",
    "label": "Module3",
    "parent": TocTreeItem {
      "children": [Circular],
      "collapsibleState": "Collapsed",
      "command": Object {
        "arguments": Array [
          Object {
            "$mid": 1,
            "path": "/tmp/fakeworkspace/collections/collection2.collection.xml",
            "scheme": "file",
          },
        ],
        "command": "vscode.open",
        "title": "open",
      },
      "description": undefined,
      "iconPath": ThemeIcon {},
      "label": "Collection2",
      "parent": undefined,
    },
  },
]
`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 5`] = `
TocTreeItem {
  "children": Array [
    TocTreeItem {
      "children": Array [],
      "collapsibleState": "None",
      "command": Object {
        "arguments": Array [
          Object {
            "$mid": 1,
            "path": "/tmp/fakeworkspace/modules/m00003/index.cnxml",
            "scheme": "file",
          },
        ],
        "command": "vscode.open",
        "title": "open",
      },
      "description": "m00003",
      "iconPath": "File",
      "label": "Module3",
      "parent": [Circular],
    },
  ],
  "collapsibleState": "Collapsed",
  "command": Object {
    "arguments": Array [
      Object {
        "$mid": 1,
        "path": "/tmp/fakeworkspace/collections/collection2.collection.xml",
        "scheme": "file",
      },
    ],
    "command": "vscode.open",
    "title": "open",
  },
  "description": undefined,
  "iconPath": ThemeIcon {},
  "label": "Collection2",
  "parent": undefined,
}
`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 6`] = `undefined`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 7`] = `
TocTreeItem {
  "children": Array [
    TocTreeItem {
      "children": Array [],
      "collapsibleState": "None",
      "command": Object {
        "arguments": Array [
          Object {
            "$mid": 1,
            "path": "/tmp/fakeworkspace/modules/m00003/index.cnxml",
            "scheme": "file",
          },
        ],
        "command": "vscode.open",
        "title": "open",
      },
      "description": "m00003",
      "iconPath": "File",
      "label": "Module3",
      "parent": [Circular],
    },
  ],
  "collapsibleState": "Collapsed",
  "command": Object {
    "arguments": Array [
      Object {
        "$mid": 1,
        "path": "/tmp/fakeworkspace/collections/collection2.collection.xml",
        "scheme": "file",
      },
    ],
    "command": "vscode.open",
    "title": "open",
  },
  "description": undefined,
  "iconPath": ThemeIcon {},
  "label": "Collection2",
  "parent": undefined,
}
`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 8`] = `
TocTreeItem {
  "children": Array [
    TocTreeItem {
      "children": Array [],
      "collapsibleState": "None",
      "command": Object {
        "arguments": Array [
          Object {
            "$mid": 1,
            "path": "/tmp/fakeworkspace/modules/m00001/index.cnxml",
            "scheme": "file",
          },
        ],
        "command": "vscode.open",
        "title": "open",
      },
      "description": "m00001",
      "iconPath": "File",
      "label": "Module1",
      "parent": [Circular],
    },
    TocTreeItem {
      "children": Array [],
      "collapsibleState": "None",
      "command": Object {
        "arguments": Array [
          Object {
            "$mid": 1,
            "path": "/tmp/fakeworkspace/modules/m00002/index.cnxml",
            "scheme": "file",
          },
        ],
        "command": "vscode.open",
        "title": "open",
      },
      "description": "m00002",
      "iconPath": "File",
      "label": "Module2",
      "parent": [Circular],
    },
  ],
  "collapsibleState": "Collapsed",
  "command": undefined,
  "description": undefined,
  "iconPath": "Folder",
  "label": "subcollection",
  "parent": TocTreeItem {
    "children": Array [
      [Circular],
    ],
    "collapsibleState": "Collapsed",
    "command": Object {
      "arguments": Array [
        Object {
          "$mid": 1,
          "path": "/tmp/fakeworkspace/collections/collection1.collection.xml",
          "scheme": "file",
        },
      ],
      "command": "vscode.open",
      "title": "open",
    },
    "description": undefined,
    "iconPath": ThemeIcon {},
    "label": "Collection1",
    "parent": undefined,
  },
}
`;

exports[`Toc Editor TocTreesProvider returns expected TocTreeItems 9`] = `
TocTreeItem {
  "children": Array [],
  "collapsibleState": "None",
  "command": Object {
    "arguments": Array [
      Object {
        "$mid": 1,
        "path": "/tmp/fakeworkspace/modules/m00003/index.cnxml",
        "scheme": "file",
      },
    ],
    "command": "vscode.open",
    "title": "open",
  },
  "description": undefined,
  "iconPath": "File",
  "label": "Module3 (m00003)",
  "parent": undefined,
}
`;

exports[`Toc Editor filtering toggleTocTreesFilteringHandler 1`] = `
Array [
  Object {
    "args": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "label": "m1",
          },
        ],
        "label": "col2",
      },
      Object {
        "expand": 3,
      },
    ],
    "callId": 17,
    "callback": undefined,
    "errorWithCallStack": [Error],
    "exception": undefined,
    "firstArg": Object {
      "children": Array [
        Object {
          "children": Array [],
          "label": "m1",
        },
      ],
      "label": "col2",
    },
    "lastArg": Object {
      "expand": 3,
    },
    "proxy": [Function],
    "returnValue": undefined,
    "thisValue": Object {
      "reveal": [Function],
    },
  },
  Object {
    "args": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "label": "m2",
              },
            ],
            "label": "subcol",
          },
        ],
        "label": "col1",
      },
      Object {
        "expand": 3,
      },
    ],
    "callId": 18,
    "callback": undefined,
    "errorWithCallStack": [Error],
    "exception": undefined,
    "firstArg": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "label": "m2",
            },
          ],
          "label": "subcol",
        },
      ],
      "label": "col1",
    },
    "lastArg": Object {
      "expand": 3,
    },
    "proxy": [Function],
    "returnValue": undefined,
    "thisValue": Object {
      "reveal": [Function],
    },
  },
]
`;
